configuration/config2/XSDSchemaParser.cpp

```diff
@@ -349,57 +385,89 @@ void XSDSchemaParser::parseElement(const pt::ptree &elemTree)
                 pCfgValue->setType(pSimpleType);                      // will throw if type is not defined
                 pNewSchemaItem->setItemSchemaValue(pCfgValue);
-                std::shared_ptr<XSDSchemaParser> pXSDParaser = std::make_shared<XSDSchemaParser>(pNewSchemaItem);
-                pXSDParaser->parseXSD(childTree);
-                m_pSchemaItem->addChild(pNewSchemaItem);
             }
             else
             {
-                //
-                // Wasn't a simple type, complex?
                 std::shared_ptr<SchemaItem> pConfigType = m_pSchemaItem->getSchemaType(typeName, false);
                 if (pConfigType != nullptr)
                 {
-                    //
-                    // A complex type was found. Insert it
-                    std::vector<std::shared_ptr<SchemaItem>> typeChildren;
-                    pConfigType->getChildren(typeChildren);
-                    for (auto childIt = typeChildren.begin(); childIt != typeChildren.end(); ++childIt)
-                    {
-                        std::shared_ptr<SchemaItem> pNewItem = std::make_shared<SchemaItem>(*(*childIt));
-                        pNewItem->setParent(m_pSchemaItem);
-                        m_pSchemaItem->addChild(pNewItem);
-                    }
-                }
-                else
-                {
-                    std::string msg = "Unable to find type " + typeName + " when parsing element " + elementName;
-                    throw(ParseException(msg));
+                    pNewSchemaItem->insertSchemaType(pConfigType);
                 }
             }
         }
+
+        //
+        // Now parse the element child tree (note if complex was inserted, this is probably empty)
+        std::shared_ptr<XSDSchemaParser> pXSDParaser = std::make_shared<XSDSchemaParser>(pNewSchemaItem);
+        pXSDParaser->parseXSD(childTree);
+        m_pSchemaItem->addChild(pNewSchemaItem);
+    }
+}
+
+
+void XSDSchemaParser::processSchemaInsert(const pt::ptree &elemTree)
+{
+    std::string path = elemTree.get("<xmlattr>.hpcc:schemaPath", "");
+
+    //
+    // Make sure path is present and well formed
+    if (path.empty() || path[0] != '/' || path.back() == '/')
+    {
+        std::string msg = "Insert schema path is missing, empty or not welformed";
+        throw(ParseException(msg));
+    }
+
+    std::string itemType = elemTree.get("<xmlattr>.hpcc:itemType", "");
+
+    std::vector<std::shared_ptr<SchemaItem>> insertChildren;
+    std::vector<std::string> pathParts = splitString(path, "/");
+    std::shared_ptr<SchemaItem> pInsertItem = m_pSchemaItem->getSchemaRoot();
+    size_t numPasses = pathParts.size() - 1;
+
+    for (size_t i=0; i<numPasses; ++i)
+    {
+        bool isLast = i == (numPasses - 1);
+        if (pInsertItem->getProperty("name") == pathParts[i])
+        {
+            insertChildren.clear();
+            pInsertItem->getChildren(insertChildren, pathParts[i+1], isLast ? itemType : "");
```

Comment:
If isLast, can you use pathParts[i+1]?

Reply:
Yes, The loop is done numPasses times which is set to be less than the actual length of the buffer by one, so indexing forward is OK.

Reply:
When isLast = true, the i will be numPasses - 1. The  pathParts[i+1] will be pathParts[numPasses] which will be invalid. ?

Reply:
numPasses is the size of pathParts - 1, so pathParts[numPasses] will still be valid.

Reply:
Are you only looping pathParts.size() - 1 times here?

Reply:
Actually, since the loop is to i<numPasses, it's only up to pathParts.size()-2. So, a reference to pathParts[i+1] is always valid.